### PR TITLE
Drupal 9 compatibility and bug fixes for other deprecated functions

### PR DESCRIPTION
**Pull request by fathimasmat**
This pull request includes commits that fix deprecated functions and other bugs on the D9 version. The fixes should consistently work on both Drupal versions 8 and 9.

### CODE REVIEW

**Comment by fathimasmat** [comment]
This should be the updated PR/version for D9 compatibility @sun . Could someone review and merge or move it to a separate branch at-least so I can use the base repo for my Drupal project ? Currently I'm using my forked repo to use this module with my updates to it.

File: src/Component.php [hunk 1]
```diff
@@ -231,6 +239,14 @@ protected function getVariantDefaultVariables(string $modifier, array $defined_v
    */
   protected function extractParts(string $compound_name): array {
     $loader = $this->env->getLoader();
+    if (stripos($compound_name, '@') !== FALSE && stripos($compound_name, '.twig') === FALSE) {
+      $exploded_paths = explode('/', $compound_name);
+      $exploded_paths = array_filter($exploded_paths);
+      $last_part = $exploded_paths[count($exploded_paths) - 1];
+      $last_part = str_replace('@', '', $last_part);
+      $compound_name .= '/' . $last_part . '.twig';
+    }
```
Comment (sun) [inline src/Component.php:240]: Why are you trying to resolve the namespaced path here? The loader is doing that already?

File: src/Node/Render.php [hunk 2]
```diff
@@ -51,7 +51,14 @@ protected function setEnvironment(\Twig_Environment $env) {
   }
 
   public static function getEnvironment(): \Twig_Environment {
-    return static::$env;
+    $env = static::$env;
+    if (!$env) {
+      $loader = \Drupal::service('twig.loader.twig_fractal');
+      $env = new \Twig\Environment($loader);
```
Comment (sun) [inline src/Node/Render.php]: Please revert this. The module is not supposed to create a new twig environment. That needs to be done in the wrapping application which uses the module/package.

File: src/Node/Render.php [hunk 3]
```diff
@@ -69,43 +76,50 @@ public static function getEnvironment(): \Twig_Environment {
    */
   protected function addGetTemplate(Twig_Compiler $compiler) {
     $this->setEnvironment($compiler->getEnvironment());
-    $compiler->raw('$handles = (array) ')->subcompile($this->getNode('expr'))->raw(';');
+    $compiler->raw('$handles = (array) ')
+      ->subcompile($this->getNode('expr'))
+      ->raw(';');
```
Comment (sun) [inline src/Node/Render.php]: Can you revert all the coding style changes, please? You can propose them in a separate PR, but for this one they are out of scope.

File: twig_fractal.services.yml [hunk 4]
```diff
@@ -2,10 +2,10 @@ services:
...
-      - { name: twig.extension }
+      - { name: twig.extension , priority: 220 }
   twig.loader.twig_fractal:
-      class: Drupal\twig_fractal\TwigLoader
-      arguments: ['@app.root', '@module_handler', '@theme_handler']
-      tags:
-        # twig.loader.componentlibrary is registered with 200
-        - { name: twig.loader, priority: 210 }
+    class: Drupal\twig_fractal\TwigLoader
+    arguments: [ '@theme_handler','@components.info', '@theme.manager' ]
```
Comment (sun) [inline twig_fractal.services.yml]: This appears to be the only change which would require a new branch/mainline for D9. Is there a way to avoid that and support both versions in a single mainline?

**Review by sun — CHANGES_REQUESTED** [review]
Thank you for looking into this! 🙂 I reviewed the changes and found at least one that needs rethinking.

I'd also appreciate your thoughts and ideas on how we could maintain a single mainline in the package (avoiding a parallel one for existing implementations). That said, we ourselves are only using this package as a stand-alone extension for custom Twig environments in WordPress and custom applications, so I personally wouldn't have a problem with simply moving to D9 in the mainline and dropping support for D8. I don't know whether anyone else is using this package.

**Comment by sun** [comment]
@fathimasmat Did you already have a chance to look into the feedback on the code?

File: src/Component.php (hunk 1 above)
Comment (fathimasmat) [reply src/Component.php:240]: Hi, the namespace is resolved when the 'render' tag includes the twig file path as well but doesn't resolve when twig path is missing.
Ex:  `{% render "@card/card.twig" %}` is working
        `{% render "@card" %}` is not working
As per Fractal standards, the latter is supposed to auto pick the "card.twig" template from the component by default. So trying to insist that behaviour by checking if .twig exists in the path and auto populate the default template if not provided as in [https://github.com/netzstrategen/twig_fractal/pull/30/files/1729f0588a258f12685f1de04e2dc72a272326b8#diff-79446737c5fc656d037e5ebaee2e336acb25c73dfff3d03856d42728276b9b32R242](https://github.com/netzstrategen/twig_fractal/pull/30/files/1729f0588a258f12685f1de04e2dc72a272326b8#diff-79446737c5fc656d037e5ebaee2e336acb25c73dfff3d03856d42728276b9b32R242)

File: src/Node/Render.php (hunk 3 above)
Comment (fathimasmat) [reply src/Node/Render.php]: Ok, thanks

File: twig_fractal.services.yml (hunk 4 above)
Comment (fathimasmat) [reply twig_fractal.services.yml]: I'm not sure if there is a way to handle both 8 and 9 from the mainline. Is there a way for you to support two versions/tags for this module and release the module to Drupal and make the install with composer that can pick the relevant module version based on Drupal version?

File: src/Node/Render.php (hunk 2 above)
Comment (fathimasmat) [reply src/Node/Render.php]: Right, this was an issue I had to pick when 'render' worked as twig 'include' tag without reinitiating the context for the included component and just using the same context on all rendered blocks. It appeared to be like an issue sorted within the module itself.

> That needs to be done in the wrapping application which uses the module/package.

Do you have a suggestion of how this can be controlled at the application level?

**Comment by fathimasmat** [comment]
@sun I have replied to your comments above. In case if I need to submit changes to the existing PR, how would you want me to do that?

File: src/Node/Render.php (hunk 2 above)
Comment (fathimasmat) [reply src/Node/Render.php]: jsut fyi more info on the the error thrown when the twig env is not reset

```
The website encountered an unexpected error. Please try again later.
TypeError: Return value of Drupal\twig_fractal\Node\Render::getEnvironment() must be an instance of Twig\Environment, null returned in Drupal\twig_fractal\Node\Render::getEnvironment() (line 56 of modules/contrib/twig_fractal/src/Node/Render.php).
Drupal\twig_fractal\Node\Render::getEnvironment() (Line: 67)
```

**Comment by fathimasmat** [comment]
Components Module has recently released [version 3](https://www.drupal.org/project/components/releases/3.0.0-alpha1) with additional new features, api hooks and dynamic namespaces can be loaded through [theme registry alter hoo](https://git.drupalcode.org/project/components/-/blob/3.x/components.api.php#L24)k without having to manually add the paths from theme.info yml file.

As per the newer version, the service info for the components module has been updated and I have pushed new [commits](https://github.com/netzstrategen/twig_fractal/pull/30/commits/505ae1270e9c6a4170dd2e2deb2cdf14717b5275) to replicate the newer changes in twig_fractal module.

**Comment by sun** [comment]
> I need to submit changes to the existing PR, how would you want me to do that?

@fathimasmat You can just add further commits to the branch of this PR - all commits will be squashed into a single when merging into the mainline later on.

File: twig_fractal.services.yml (hunk 4 above)
Comment (sun) [reply twig_fractal.services.yml]: No, that would require more than one main branch.  But nevermind.  Let‘s just bump the major version with this PR and it should be fine.  There were no formal releases anyway.

Unless my information is outdated, drupal.org does not support repositories hosted on GitHub yet.  And the GitHub package registry does not support Composer yet.  So until either of these two conditions changes you'd need to continue referencing the repository separately in your composer.json.

File: src/Node/Render.php (hunk 2 above)
Comment (sun) [reply src/Node/Render.php]: Sorry, I was confused here.  The Render tag does this on its own when compiling the template: https://github.com/netzstrategen/twig_fractal/blob/bf355130e411953924e87f3ee064e131de5055b4/src/Node/Render.php#L71

File: src/Component.php (hunk 1 above)
Comment (sun) [reply src/Component.php:240]: Ah okay.  In that case, it's fine; we can just simplify the code.

File: src/Component.php [hunk 5]
```diff
@@ -231,6 +239,14 @@ protected function getVariantDefaultVariables(string $modifier, array $defined_v
...
+
```
Comment (sun) [inline src/Component.php]:
```suggestion
    // Automatically expand a component name to its template filename; e.g.:
    // "@card" => "@card/card.twig"
    if (strpos($compound_name, '@') !== FALSE && stripos($compound_name, '.twig') === FALSE) {
      $parts = explode('/', $compound_name);
      $compound_name .= '/' . ltrim(end($parts), '@') . '.twig';
    }
```

File: src/Component.php (hunk 1 above)
Comment (fathimasmat) [reply src/Component.php:240]: thanks, this change is added now

File: twig_fractal.services.yml (hunk 4 above)
Comment (fathimasmat) [reply twig_fractal.services.yml]: the services yml was updated last week to account he new changes to the components module - https://github.com/netzstrategen/twig_fractal/pull/30#issuecomment-773347515

File: src/Node/Render.php (hunk 2 above)
Comment (fathimasmat) [reply src/Node/Render.php]: for some reason, `$compiler->getEnvironment()` seems to be null after first load. When the pages with patterns are loaded first time in Drupal, twig is loaded fine and rendering the patterns as expected but in the second load, it throws up the error that I mentioned above stating` Drupal\twig_fractal\Node\Render::getEnvironment() must be an instance of Twig\Environment, null returned`. 
After clearing cache from Drupal then Twig env is back and again in the second load being null.

should there be a cache handler setup to store the Twig env state?

File: src/Node/Render.php (hunk 3 above)
Comment (fathimasmat) [reply src/Node/Render.php]: @sun coding standards changes are reverted now https://github.com/fathimasmat/twig_fractal/commit/6802008b4fe713ab480c2ba34c70e2d73eb32842 and https://github.com/fathimasmat/twig_fractal/commit/8d06f3f0f12384f484313fd2782725da03629c3a

**Comment by fathimasmat** [comment]
@sun I replied to all the comments and resolved changes as required, please review and merge if possible

File: src/Node/Render.php (hunk 2 above)
Comment (fathimasmat) [reply src/Node/Render.php]: @sun just realised this has been already added as an issue by another user here - https://github.com/netzstrategen/twig_fractal/issues/26

File: twig_fractal.services.yml (hunk 4 above)
Comment (fathimasmat) [reply twig_fractal.services.yml]: Hello @sun did you have a chance to look at my further comments above and recent commits to the PR? Really looking forward to get this one merged over to the main line!

**Comment by dasginganinja** [comment]
@sun Is there an update on this issue? We are looking to use this with Drupal 9.

**Comment by sun** [comment]
Sorry for the delay here.  Unfortunately the PR never got updated to revert the unrelated changes, so they are still contained here today.

If someone wants to resurrect this feature, I believe the best way forward is to create a new PR that only contains the necessary functional changes for the integration with Drupal's new Components registry.

That would mean the changes visible here: https://github.com/netzstrategen/twig_fractal/pull/30/files?diff=unified&w=1 (minus the first line/whitespace change).  Everything else should be left untouched.  I think this would allow us to move forward with the PR fairly quickly. 👍 

Thanks everyone!
